// @azure/keyvault-keys/dist-esm/keyvault-keys/src/index.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
/* eslint @typescript-eslint/member-ordering: 0 */
/// <reference lib="esnext.asynciterable" />
import { __asyncGenerator, __asyncValues, __await, __awaiter, __rest } from "tslib";
import { createPipelineFromOptions, isTokenCredential, operationOptionsToRequestOptionsBase, signingPolicy } from "@azure/core-http";
import { getTracer } from "@azure/core-tracing";
import { logger } from "./log";
import "@azure/core-paging";
import { KeyVaultClient } from "./generated/keyVaultClient";
import { SDK_VERSION } from "./generated/utils/constants";
import { challengeBasedAuthenticationPolicy } from "../../keyvault-common/src";
import { DeleteKeyPoller } from "./lro/delete/poller";
import { RecoverDeletedKeyPoller } from "./lro/recover/poller";
import { LATEST_API_VERSION } from "./keysModels";
import { parseKeyvaultIdentifier as parseKeyvaultEntityIdentifier } from "./generated/utils";
import { CryptographyClient } from "./cryptographyClient";
export { CryptographyClient, logger };
/**
 * The KeyClient provides methods to manage {@link KeyVaultKey} in the
 * Azure Key Vault. The client supports creating, retrieving, updating,
 * deleting, purging, backing up, restoring and listing KeyVaultKeys. The
 * client also supports listing {@link DeletedKey} for a soft-delete enabled Azure Key
 * Vault.
 */
export class KeyClient {
    /**
     * Creates an instance of KeyClient.
     *
     * Example usage:
     * ```ts
     * import { KeyClient } from "@azure/keyvault-keys";
     * import { DefaultAzureCredential } from "@azure/identity";
     *
     * let vaultUrl = `https://<MY KEYVAULT HERE>.vault.azure.net`;
     * let credentials = new DefaultAzureCredential();
     *
     * let client = new KeyClient(vaultUrl, credentials);
     * ```
     * @param {string} vaultUrl the URL of the Key Vault. It should have this shape: https://${your-key-vault-name}.vault.azure.net
     * @param {TokenCredential} credential An object that implements the `TokenCredential` interface used to authenticate requests to the service. Use the @azure/identity package to create a credential that suits your needs.
     * @param {PipelineOptions} [pipelineOptions] Pipeline options used to configure Key Vault API requests. Omit this parameter to use the default pipeline configuration.
     * @memberof KeyClient
     */
    constructor(vaultUrl, credential, pipelineOptions = {}) {
        /**
         * @internal
         * @ignore
         * A self reference that bypasses private methods, for the pollers.
         */
        this.pollerClient = {
            recoverDeletedKey: this.recoverDeletedKey.bind(this),
            getKey: this.getKey.bind(this),
            deleteKey: this.deleteKey.bind(this),
            getDeletedKey: this.getDeletedKey.bind(this)
        };
        this.vaultUrl = vaultUrl;
        const libInfo = `azsdk-js-keyvault-keys/${SDK_VERSION}`;
        const userAgentOptions = pipelineOptions.userAgentOptions;
        pipelineOptions.userAgentOptions = Object.assign(Object.assign({}, pipelineOptions.userAgentOptions), { userAgentPrefix: userAgentOptions && userAgentOptions.userAgentPrefix
                ? `${userAgentOptions.userAgentPrefix} ${libInfo}`
                : libInfo });
        const authPolicy = isTokenCredential(credential)
            ? challengeBasedAuthenticationPolicy(credential)
            : signingPolicy(credential);
        const internalPipelineOptions = Object.assign(Object.assign({}, pipelineOptions), {
            loggingOptions: {
                logger: logger.info,
                logPolicyOptions: {
                    allowedHeaderNames: [
                        "x-ms-keyvault-region",
                        "x-ms-keyvault-network-info",
                        "x-ms-keyvault-service-version"
                    ]
                }
            }
        });
        const pipeline = createPipelineFromOptions(internalPipelineOptions, authPolicy);
        this.client = new KeyVaultClient(pipelineOptions.serviceVersion || LATEST_API_VERSION, pipeline);
    }
    /**
     * @internal
     * @ignore
     * Sends a delete request for the given KeyVault Key's name to the KeyVault service.
     * Since the KeyVault Key won't be immediately deleted, we have {@link beginDeleteKey}.
     * @param {string} name The name of the KeyVault Key.
     * @param {DeleteKeyOptions} [options] Optional parameters for the underlying HTTP request.
     */
    deleteKey(name, options = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            const requestOptions = operationOptionsToRequestOptionsBase(options);
            const span = this.createSpan("deleteKey", requestOptions);
            let response;
            try {
                response = yield this.client.deleteKey(this.vaultUrl, name, this.setParentSpan(span, requestOptions));
            }
            finally {
                span.end();
            }
            return this.getKeyFromKeyBundle(response);
        });
    }
    /**
     * @internal
     * @ignore
     * Sends a request to recover a deleted KeyVault Key based on the given name.
     * Since the KeyVault Key won't be immediately recover the deleted key, we have {@link beginRecoverDeletedKey}.
     * @param {string} name The name of the KeyVault Key.
     * @param {RecoverDeletedKeyOptions} [options] Optional parameters for the underlying HTTP request.
     */
    recoverDeletedKey(name, options = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            const requestOptions = operationOptionsToRequestOptionsBase(options);
            const span = this.createSpan("recoverDeletedKey", requestOptions);
            let response;
            try {
                response = yield this.client.recoverDeletedKey(this.vaultUrl, name, this.setParentSpan(span, requestOptions));
            }
            finally {
                span.end();
            }
            return this.getKeyFromKeyBundle(response);
        });
    }
    /**
     * The create key operation can be used to create any key type in Azure Key Vault. If the named key
     * already exists, Azure Key Vault creates a new version of the key. It requires the keys/create
     * permission.
     *
     * Example usage:
     * ```ts
     * let client = new KeyClient(url, credentials);
     * // Create an elliptic-curve key:
     * let result = await client.createKey("MyKey", "EC");
     * ```
     * @summary Creates a new key, stores it, then returns key parameters and properties to the client.
     * @param {string} name The name of the key.
     * @param {KeyType} keyType The type of the key. One of the following: 'EC', 'EC-HSM', 'RSA', 'RSA-HSM', 'oct'.
     * @param {CreateKeyOptions} [options] The optional parameters.
     */
    createKey(name, keyType, options) {
        return __awaiter(this, void 0, void 0, function* () {
            if (options) {
                const requestOptions = operationOptionsToRequestOptionsBase(options);
                const { enabled, notBefore, expiresOn: expires } = requestOptions, remainingOptions = __rest(requestOptions, ["enabled", "notBefore", "expiresOn"]);
                const unflattenedOptions = Object.assign(Object.assign({}, remainingOptions), { keyAttributes: {
                        enabled,
                        notBefore,
                        expires
                    } });
                const span = this.createSpan("createKey", unflattenedOptions);
                let response;
                try {
                    response = yield this.client.createKey(this.vaultUrl, name, keyType, this.setParentSpan(span, unflattenedOptions));
                }
                finally {
                    span.end();
                }
                return this.getKeyFromKeyBundle(response);
            }
            else {
                const response = yield this.client.createKey(this.vaultUrl, name, keyType, options);
                return this.getKeyFromKeyBundle(response);
            }
        });
    }
    /**
     * The createEcKey method creates a new elliptic curve key in Azure Key Vault. If the named key
     * already exists, Azure Key Vault creates a new version of the key. It requires the keys/create
     * permission.
     *
     * Example usage:
     * ```ts
     * let client = new KeyClient(url, credentials);
     * let result = await client.createEcKey("MyKey", { curve: "P-256" });
     * ```
     * @summary Creates a new key, stores it, then returns key parameters and properties to the client.
     * @param {string} name The name of the key.
     * @param {CreateEcKeyOptions} [options] The optional parameters.
     */
    createEcKey(name, options) {
        return __awaiter(this, void 0, void 0, function* () {
            if (options) {
                const requestOptions = operationOptionsToRequestOptionsBase(options);
                const { enabled, notBefore, expiresOn: expires } = requestOptions, remainingOptions = __rest(requestOptions, ["enabled", "notBefore", "expiresOn"]);
                const unflattenedOptions = Object.assign(Object.assign({}, remainingOptions), { keyAttributes: {
                        enabled,
                        notBefore,
                        expires
                    } });
                const span = this.createSpan("createEcKey", unflattenedOptions);
                let response;
                try {
                    response = yield this.client.createKey(this.vaultUrl, name, options.hsm ? "EC-HSM" : "EC", this.setParentSpan(span, unflattenedOptions));
                }
                finally {
                    span.end();
                }
                return this.getKeyFromKeyBundle(response);
            }
            else {
                const response = yield this.client.createKey(this.vaultUrl, name, "EC", options);
                return this.getKeyFromKeyBundle(response);
            }
        });
    }
    /**
     * The createRSAKey method creates a new RSA key in Azure Key Vault. If the named key
     * already exists, Azure Key Vault creates a new version of the key. It requires the keys/create
     * permission.
     *
     * Example usage:
     * ```ts
     * let client = new KeyClient(url, credentials);
     * let result = await client.createRsaKey("MyKey", { keySize: 2048 });
     * ```
     * @summary Creates a new key, stores it, then returns key parameters and properties to the client.
     * @param {string} name The name of the key.
     * @param {CreateRsaKeyOptions} [options] The optional parameters.
     */
    createRsaKey(name, options) {
        return __awaiter(this, void 0, void 0, function* () {
            if (options) {
                const requestOptions = operationOptionsToRequestOptionsBase(options);
                const { enabled, notBefore, expiresOn: expires } = requestOptions, remainingOptions = __rest(requestOptions, ["enabled", "notBefore", "expiresOn"]);
                const unflattenedOptions = Object.assign(Object.assign({}, remainingOptions), { keyAttributes: {
                        enabled,
                        notBefore,
                        expires
                    } });
                const span = this.createSpan("createRsaKey", unflattenedOptions);
                let response;
                try {
                    response = yield this.client.createKey(this.vaultUrl, name, options.hsm ? "RSA-HSM" : "RSA", this.setParentSpan(span, unflattenedOptions));
                }
                finally {
                    span.end();
                }
                return this.getKeyFromKeyBundle(response);
            }
            else {
                const response = yield this.client.createKey(this.vaultUrl, name, "RSA", options);
                return this.getKeyFromKeyBundle(response);
            }
        });
    }
    /**
     * The import key operation may be used to import any key type into an Azure Key Vault. If the
     * named key already exists, Azure Key Vault creates a new version of the key. This operation
     * requires the keys/import permission.
     *
     * Example usage:
     * ```ts
     * let client = new KeyClient(url, credentials);
     * // Key contents in myKeyContents
     * let result = await client.importKey("MyKey", myKeyContents);
     * ```
     * @summary Imports an externally created key, stores it, and returns key parameters and properties
     * to the client.
     * @param {string} name Name for the imported key.
     * @param {JsonWebKey} key The JSON web key.
     * @param {ImportKeyOptions} [options] The optional parameters.
     */
    importKey(name, key, options) {
        return __awaiter(this, void 0, void 0, function* () {
            if (options) {
                const requestOptions = operationOptionsToRequestOptionsBase(options);
                const { enabled, notBefore, expiresOn: expires, hardwareProtected: hsm } = requestOptions, remainingOptions = __rest(requestOptions, ["enabled", "notBefore", "expiresOn", "hardwareProtected"]);
                const unflattenedOptions = Object.assign(Object.assign({}, remainingOptions), { keyAttributes: {
                        enabled,
                        notBefore,
                        expires,
                        hsm
                    } });
                const span = this.createSpan("importKey", unflattenedOptions);
                let response;
                try {
                    response = yield this.client.importKey(this.vaultUrl, name, key, this.setParentSpan(span, unflattenedOptions));
                }
                finally {
                    span.end();
                }
                return this.getKeyFromKeyBundle(response);
            }
            else {
                const response = yield this.client.importKey(this.vaultUrl, name, key, options);
                return this.getKeyFromKeyBundle(response);
            }
        });
    }
    /**
     * The delete operation applies to any key stored in Azure Key Vault. Individual versions
     * of a key can not be deleted, only all versions of a given key at once.
     *
     * This function returns a Long Running Operation poller that allows you to wait indefinitely until the key is deleted.
     *
     * This operation requires the keys/delete permission.
     *
     * Example usage:
     * ```ts
     * const client = new KeyClient(url, credentials);
     * await client.createKey("MyKey", "EC");
     * const poller = await client.beginDeleteKey("MyKey");
     *
     * // Serializing the poller
     * const serialized = poller.toString();
     * // A new poller can be created with:
     * // await client.beginDeleteKey("MyKey", { resumeFrom: serialized });
     *
     * // Waiting until it's done
     * const deletedKey = await poller.pollUntilDone();
     * console.log(deletedKey);
     * ```
     * @summary Deletes a key from a specified key vault.
     * @param {string} name The name of the key.
     * @param {BeginDeleteKeyOptions} [options] The optional parameters.
     */
    beginDeleteKey(name, options = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            const requestOptions = operationOptionsToRequestOptionsBase(options);
            const poller = new DeleteKeyPoller({
                name,
                client: this.pollerClient,
                intervalInMs: options.intervalInMs,
                resumeFrom: options.resumeFrom,
                requestOptions
            });
            // This will initialize the poller's operation (the deletion of the key).
            yield poller.poll();
            return poller;
        });
    }
    /**
     * The updateKeyProperties method changes specified properties of an existing stored key. Properties that
     * are not specified in the request are left unchanged. The value of a key itself cannot be
     * changed. This operation requires the keys/set permission.
     *
     * Example usage:
     * ```ts
     * let keyName = "MyKey";
     * let client = new KeyClient(url, credentials);
     * let key = await client.getKey(keyName);
     * let result = await client.updateKeyProperties(keyName, key.properties.version, { enabled: false });
     * ```
     * @summary Updates the properties associated with a specified key in a given key vault.
     * @param {string} name The name of the key.
     * @param {string} keyVersion The version of the key.
     * @param {UpdateKeyPropertiesOptions} [options] The optional parameters.
     */
    updateKeyProperties(name, keyVersion, options) {
        return __awaiter(this, void 0, void 0, function* () {
            if (options) {
                const requestOptions = operationOptionsToRequestOptionsBase(options);
                const { enabled, notBefore, expiresOn: expires } = requestOptions, remainingOptions = __rest(requestOptions, ["enabled", "notBefore", "expiresOn"]);
                const unflattenedOptions = Object.assign(Object.assign({}, remainingOptions), { keyAttributes: {
                        enabled,
                        notBefore,
                        expires
                    } });
                const span = this.createSpan("updateKeyProperties", unflattenedOptions);
                let response;
                try {
                    response = yield this.client.updateKey(this.vaultUrl, name, keyVersion, this.setParentSpan(span, unflattenedOptions));
                }
                finally {
                    span.end();
                }
                return this.getKeyFromKeyBundle(response);
            }
            else {
                const response = yield this.client.updateKey(this.vaultUrl, name, keyVersion, options);
                return this.getKeyFromKeyBundle(response);
            }
        });
    }
    /**
     * The getKey method gets a specified key and is applicable to any key stored in Azure Key Vault.
     * This operation requires the keys/get permission.
     *
     * Example usage:
     * ```ts
     * let client = new KeyClient(url, credentials);
     * let key = await client.getKey("MyKey");
     * ```
     * @summary Get a specified key from a given key vault.
     * @param {string} name The name of the key.
     * @param {GetKeyOptions} [options] The optional parameters.
     */
    getKey(name, options = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            const requestOptions = operationOptionsToRequestOptionsBase(options);
            const span = this.createSpan("getKey", requestOptions);
            let response;
            try {
                response = yield this.client.getKey(this.vaultUrl, name, options && options.version ? options.version : "", this.setParentSpan(span, requestOptions));
            }
            finally {
                span.end();
            }
            return this.getKeyFromKeyBundle(response);
        });
    }
    /**
     * The getDeletedKey method returns the specified deleted key along with its properties.
     * This operation requires the keys/get permission.
     *
     * Example usage:
     * ```ts
     * let client = new KeyClient(url, credentials);
     * let key = await client.getDeletedKey("MyDeletedKey");
     * ```
     * @summary Gets the specified deleted key.
     * @param {string} name The name of the key.
     * @param {GetDeletedKeyOptions} [options] The optional parameters.
     */
    getDeletedKey(name, options = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            const responseOptions = operationOptionsToRequestOptionsBase(options);
            const span = this.createSpan("getDeletedKey", responseOptions);
            let response;
            try {
                response = yield this.client.getDeletedKey(this.vaultUrl, name, this.setParentSpan(span, responseOptions));
            }
            finally {
                span.end();
            }
            return this.getKeyFromKeyBundle(response);
        });
    }
    /**
     * The purge deleted key operation removes the key permanently, without the possibility of
     * recovery. This operation can only be enabled on a soft-delete enabled vault. This operation
     * requires the keys/purge permission.
     *
     * Example usage:
     * ```ts
     * const client = new KeyClient(url, credentials);
     * const deletePoller = await client.beginDeleteKey("MyKey")
     * await deletePoller.pollUntilDone();
     * await client.purgeDeletedKey("MyKey");
     * ```
     * @summary Permanently deletes the specified key.
     * @param name The name of the key.
     * @param {PurgeDeletedKeyOptions} [options] The optional parameters.
     */
    purgeDeletedKey(name, options = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            const responseOptions = operationOptionsToRequestOptionsBase(options);
            const span = this.createSpan("purgeDeletedKey", responseOptions);
            try {
                yield this.client.purgeDeletedKey(this.vaultUrl, name, this.setParentSpan(span, responseOptions));
            }
            finally {
                span.end();
            }
        });
    }
    /**
     * Recovers the deleted key in the specified vault. This operation can only be performed on a
     * soft-delete enabled vault.
     *
     * This function returns a Long Running Operation poller that allows you to wait indefinitely until the deleted key is recovered.
     *
     * This operation requires the keys/recover permission.
     *
     * Example usage:
     * ```ts
     * const client = new KeyClient(url, credentials);
     * await client.createKey("MyKey", "EC");
     * const deletePoller = await client.beginDeleteKey("MyKey");
     * await deletePoller.pollUntilDone();
     * const poller = await client.beginRecoverDeletedKey("MyKey");
     *
     * // Serializing the poller
     * const serialized = poller.toString();
     * // A new poller can be created with:
     * // await client.beginRecoverDeletedKey("MyKey", { resumeFrom: serialized });
     *
     * // Waiting until it's done
     * const key = await poller.pollUntilDone();
     * console.log(key);
     * ```
     * @summary Recovers the deleted key to the latest version.
     * @param name The name of the deleted key.
     * @param {BeginRecoverDeletedKeyOptions} [options] The optional parameters.
     */
    beginRecoverDeletedKey(name, options = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            const requestOptions = operationOptionsToRequestOptionsBase(options);
            const poller = new RecoverDeletedKeyPoller({
                name,
                client: this.pollerClient,
                intervalInMs: options.intervalInMs,
                resumeFrom: options.resumeFrom,
                requestOptions: requestOptions
            });
            // This will initialize the poller's operation (the deletion of the key).
            yield poller.poll();
            return poller;
        });
    }
    /**
     * Requests that a backup of the specified key be downloaded to the client. All versions of the
     * key will be downloaded. This operation requires the keys/backup permission.
     *
     * Example usage:
     * ```ts
     * let client = new KeyClient(url, credentials);
     * let backupContents = await client.backupKey("MyKey");
     * ```
     * @summary Backs up the specified key.
     * @param {string} name The name of the key.
     * @param {BackupKeyOptions} [options] The optional parameters.
     */
    backupKey(name, options = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            const requestOptions = operationOptionsToRequestOptionsBase(options);
            const span = this.createSpan("backupKey", requestOptions);
            let response;
            try {
                response = yield this.client.backupKey(this.vaultUrl, name, this.setParentSpan(span, requestOptions));
            }
            finally {
                span.end();
            }
            return response.value;
        });
    }
    /**
     * Restores a backed up key, and all its versions, to a vault. This operation requires the
     * keys/restore permission.
     *
     * Example usage:
     * ```ts
     * let client = new KeyClient(url, credentials);
     * let backupContents = await client.backupKey("MyKey");
     * // ...
     * let key = await client.restoreKeyBackup(backupContents);
     * ```
     * @summary Restores a backed up key to a vault.
     * @param {Uint8Array} backup The backup blob associated with a key bundle.
     * @param {RestoreKeyBackupOptions} [options] The optional parameters.
     */
    restoreKeyBackup(backup, options = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            const requestOptions = operationOptionsToRequestOptionsBase(options);
            const span = this.createSpan("restoreKeyBackup", requestOptions);
            let response;
            try {
                response = yield this.client.restoreKey(this.vaultUrl, backup, this.setParentSpan(span, requestOptions));
            }
            finally {
                span.end();
            }
            return this.getKeyFromKeyBundle(response);
        });
    }
    /**
     * @internal
     * @ignore
     * Deals with the pagination of {@link listPropertiesOfKeyVersions}.
     * @param {string} name The name of the KeyVault Key.
     * @param {PageSettings} continuationState An object that indicates the position of the paginated request.
     * @param {ListPropertiesOfKeyVersionsOptions} [options] Common options for the iterative endpoints.
     */
    listPropertiesOfKeyVersionsPage(name, continuationState, options) {
        return __asyncGenerator(this, arguments, function* listPropertiesOfKeyVersionsPage_1() {
            if (continuationState.continuationToken == null) {
                const optionsComplete = Object.assign({ maxresults: continuationState.maxPageSize }, options);
                const currentSetResponse = yield __await(this.client.getKeyVersions(this.vaultUrl, name, optionsComplete));
                continuationState.continuationToken = currentSetResponse.nextLink;
                if (currentSetResponse.value) {
                    yield yield __await(currentSetResponse.value.map(this.getKeyPropertiesFromKeyItem));
                }
            }
            while (continuationState.continuationToken) {
                const currentSetResponse = yield __await(this.client.getKeyVersions(continuationState.continuationToken, name, options));
                continuationState.continuationToken = currentSetResponse.nextLink;
                if (currentSetResponse.value) {
                    yield yield __await(currentSetResponse.value.map(this.getKeyPropertiesFromKeyItem));
                }
                else {
                    break;
                }
            }
        });
    }
    /**
     * @internal
     * @ignore
     * Deals with the iteration of all the available results of {@link listPropertiesOfKeyVersions}.
     * @param {string} name The name of the KeyVault Key.
     * @param {ListPropertiesOfKeyVersionsOptions} [options] Common options for the iterative endpoints.
     */
    listPropertiesOfKeyVersionsAll(name, options) {
        return __asyncGenerator(this, arguments, function* listPropertiesOfKeyVersionsAll_1() {
            var e_1, _a;
            const f = {};
            try {
                for (var _b = __asyncValues(this.listPropertiesOfKeyVersionsPage(name, f, options)), _c; _c = yield __await(_b.next()), !_c.done;) {
                    const page = _c.value;
                    for (const item of page) {
                        yield yield __await(item);
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) yield __await(_a.call(_b));
                }
                finally { if (e_1) throw e_1.error; }
            }
        });
    }
    /**
     * Iterates all versions of the given key in the vault. The full key identifier, properties, and tags are provided
     * in the response. This operation requires the keys/list permission.
     *
     * Example usage:
     * ```ts
     * let client = new KeyClient(url, credentials);
     * for await (const keyProperties of client.listPropertiesOfKeyVersions("MyKey")) {
     *   const key = await client.getKey(keyProperties.name);
     *   console.log("key version: ", key);
     * }
     * ```
     * @param {string} name Name of the key to fetch versions for
     * @param {ListPropertiesOfKeyVersionsOptions} [options] The optional parameters.
     */
    listPropertiesOfKeyVersions(name, options = {}) {
        const requestOptions = operationOptionsToRequestOptionsBase(options);
        const span = this.createSpan("listPropertiesOfKeyVersions", requestOptions);
        const updatedOptions = Object.assign(Object.assign({}, requestOptions), this.setParentSpan(span, requestOptions));
        const iter = this.listPropertiesOfKeyVersionsAll(name, updatedOptions);
        span.end();
        return {
            next() {
                return iter.next();
            },
            [Symbol.asyncIterator]() {
                return this;
            },
            byPage: (settings = {}) => this.listPropertiesOfKeyVersionsPage(name, settings, updatedOptions)
        };
    }
    /**
     * @internal
     * @ignore
     * Deals with the pagination of {@link listPropertiesOfKeys}.
     * @param {PageSettings} continuationState An object that indicates the position of the paginated request.
     * @param {ListPropertiesOfKeysOptions} [options] Common options for the iterative endpoints.
     */
    listPropertiesOfKeysPage(continuationState, options) {
        return __asyncGenerator(this, arguments, function* listPropertiesOfKeysPage_1() {
            if (continuationState.continuationToken == null) {
                const optionsComplete = Object.assign({ maxresults: continuationState.maxPageSize }, options);
                const currentSetResponse = yield __await(this.client.getKeys(this.vaultUrl, optionsComplete));
                continuationState.continuationToken = currentSetResponse.nextLink;
                if (currentSetResponse.value) {
                    yield yield __await(currentSetResponse.value.map(this.getKeyPropertiesFromKeyItem));
                }
            }
            while (continuationState.continuationToken) {
                const currentSetResponse = yield __await(this.client.getKeys(continuationState.continuationToken, options));
                continuationState.continuationToken = currentSetResponse.nextLink;
                if (currentSetResponse.value) {
                    yield yield __await(currentSetResponse.value.map(this.getKeyPropertiesFromKeyItem));
                }
                else {
                    break;
                }
            }
        });
    }
    /**
     * @internal
     * @ignore
     * Deals with the iteration of all the available results of {@link listPropertiesOfKeys}.
     * @param {ListPropertiesOfKeysOptions} [options] Common options for the iterative endpoints.
     */
    listPropertiesOfKeysAll(options) {
        return __asyncGenerator(this, arguments, function* listPropertiesOfKeysAll_1() {
            var e_2, _a;
            const f = {};
            try {
                for (var _b = __asyncValues(this.listPropertiesOfKeysPage(f, options)), _c; _c = yield __await(_b.next()), !_c.done;) {
                    const page = _c.value;
                    for (const item of page) {
                        yield yield __await(item);
                    }
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) yield __await(_a.call(_b));
                }
                finally { if (e_2) throw e_2.error; }
            }
        });
    }
    /**
     * Iterates the latest version of all keys in the vault.  The full key identifier and properties are provided
     * in the response. No values are returned for the keys. This operations requires the keys/list permission.
     *
     * Example usage:
     * ```ts
     * let client = new KeyClient(url, credentials);
     * for await (const keyProperties of client.listPropertiesOfKeys()) {
     *   const key = await client.getKey(keyProperties.name);
     *   console.log("key: ", key);
     * }
     * ```
     * @summary List all keys in the vault
     * @param {ListPropertiesOfKeysOptions} [options] The optional parameters.
     */
    listPropertiesOfKeys(options = {}) {
        const requestOptions = operationOptionsToRequestOptionsBase(options);
        const span = this.createSpan("listPropertiesOfKeys", requestOptions);
        const updatedOptions = Object.assign(Object.assign({}, requestOptions), this.setParentSpan(span, requestOptions));
        const iter = this.listPropertiesOfKeysAll(updatedOptions);
        span.end();
        return {
            next() {
                return iter.next();
            },
            [Symbol.asyncIterator]() {
                return this;
            },
            byPage: (settings = {}) => this.listPropertiesOfKeysPage(settings, updatedOptions)
        };
    }
    /**
     * @internal
     * @ignore
     * Deals with the pagination of {@link listDeletedKeys}.
     * @param {PageSettings} continuationState An object that indicates the position of the paginated request.
     * @param {ListDeletedKeysOptions} [options] Common options for the iterative endpoints.
     */
    listDeletedKeysPage(continuationState, options) {
        return __asyncGenerator(this, arguments, function* listDeletedKeysPage_1() {
            if (continuationState.continuationToken == null) {
                const optionsComplete = Object.assign({ maxresults: continuationState.maxPageSize }, options);
                const currentSetResponse = yield __await(this.client.getDeletedKeys(this.vaultUrl, optionsComplete));
                continuationState.continuationToken = currentSetResponse.nextLink;
                if (currentSetResponse.value) {
                    yield yield __await(currentSetResponse.value.map(this.getDeletedKeyFromKeyItem));
                }
            }
            while (continuationState.continuationToken) {
                const currentSetResponse = yield __await(this.client.getDeletedKeys(continuationState.continuationToken, options));
                continuationState.continuationToken = currentSetResponse.nextLink;
                if (currentSetResponse.value) {
                    yield yield __await(currentSetResponse.value.map(this.getDeletedKeyFromKeyItem));
                }
                else {
                    break;
                }
            }
        });
    }
    /**
     * @internal
     * @ignore
     * Deals with the iteration of all the available results of {@link listDeletedKeys}.
     * @param {ListDeletedKeysOptions} [options] Common options for the iterative endpoints.
     */
    listDeletedKeysAll(options) {
        return __asyncGenerator(this, arguments, function* listDeletedKeysAll_1() {
            var e_3, _a;
            const f = {};
            try {
                for (var _b = __asyncValues(this.listDeletedKeysPage(f, options)), _c; _c = yield __await(_b.next()), !_c.done;) {
                    const page = _c.value;
                    for (const item of page) {
                        yield yield __await(item);
                    }
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) yield __await(_a.call(_b));
                }
                finally { if (e_3) throw e_3.error; }
            }
        });
    }
    /**
     * Iterates the deleted keys in the vault.  The full key identifier and properties are provided
     * in the response. No values are returned for the keys. This operations requires the keys/list permission.
     *
     * Example usage:
     * ```ts
     * let client = new KeyClient(url, credentials);
     * for await (const deletedKey of client.listDeletedKeys()) {
     *   console.log("deleted key: ", deletedKey);
     * }
     * ```
     * @summary List all keys in the vault
     * @param {ListDeletedKeysOptions} [options] The optional parameters.
     */
    listDeletedKeys(options = {}) {
        const requestOptions = operationOptionsToRequestOptionsBase(options);
        const span = this.createSpan("listDeletedKeys", requestOptions);
        const updatedOptions = Object.assign(Object.assign({}, options), this.setParentSpan(span, requestOptions));
        const iter = this.listDeletedKeysAll(updatedOptions);
        span.end();
        return {
            next() {
                return iter.next();
            },
            [Symbol.asyncIterator]() {
                return this;
            },
            byPage: (settings = {}) => this.listDeletedKeysPage(settings, updatedOptions)
        };
    }
    /**
     * @internal
     * @ignore
     * Shapes the exposed {@link KeyVaultKey} based on either a received key bundle or deleted key bundle.
     */
    getKeyFromKeyBundle(bundle) {
        const keyBundle = bundle;
        const deletedKeyBundle = bundle;
        const parsedId = parseKeyvaultEntityIdentifier("keys", keyBundle.key ? keyBundle.key.kid : undefined);
        const attributes = keyBundle.attributes || {};
        delete keyBundle.attributes;
        const resultObject = {
            key: keyBundle.key,
            id: keyBundle.key ? keyBundle.key.kid : undefined,
            name: parsedId.name,
            keyOperations: keyBundle.key ? keyBundle.key.keyOps : undefined,
            keyType: keyBundle.key ? keyBundle.key.kty : undefined,
            properties: Object.assign(Object.assign(Object.assign({ id: keyBundle.key ? keyBundle.key.kid : undefined, expiresOn: attributes.expires, createdOn: attributes.created, updatedOn: attributes.updated }, keyBundle), parsedId), attributes)
        };
        if (deletedKeyBundle.deletedDate) {
            resultObject.properties.deletedOn = deletedKeyBundle.deletedDate;
            delete resultObject.properties.deletedDate;
        }
        if (attributes.vaultUrl) {
            delete resultObject.properties.vaultUrl;
        }
        if (attributes.expires) {
            delete resultObject.properties.expires;
        }
        if (attributes.created) {
            delete resultObject.properties.created;
        }
        if (attributes.updated) {
            delete resultObject.properties.updated;
        }
        return resultObject;
    }
    /**
     * @internal
     * @ignore
     * Shapes the exposed {@link DeletedKey} based on a received KeyItem.
     */
    getDeletedKeyFromKeyItem(keyItem) {
        const parsedId = parseKeyvaultEntityIdentifier("keys", keyItem.kid);
        const attributes = keyItem.attributes || {};
        const abstractProperties = Object.assign(Object.assign(Object.assign({ id: keyItem.kid, deletedOn: attributes.deletedDate, expiresOn: attributes.expires, createdOn: attributes.created, updatedOn: attributes.updated }, keyItem), parsedId), keyItem.attributes);
        if (abstractProperties.deletedDate) {
            delete abstractProperties.deletedDate;
        }
        if (abstractProperties.expires) {
            delete abstractProperties.expires;
        }
        if (abstractProperties.created) {
            delete abstractProperties.created;
        }
        if (abstractProperties.updated) {
            delete abstractProperties.updated;
        }
        return {
            key: keyItem,
            id: keyItem.kid,
            name: abstractProperties.name,
            properties: abstractProperties
        };
    }
    /**
     * @internal
     * @ignore
     * Shapes the exposed {@link KeyProperties} based on a received KeyItem.
     */
    getKeyPropertiesFromKeyItem(keyItem) {
        const parsedId = parseKeyvaultEntityIdentifier("keys", keyItem.kid);
        const attributes = keyItem.attributes || {};
        const resultObject = Object.assign(Object.assign(Object.assign({ createdOn: attributes.created, updatedOn: attributes.updated }, keyItem), parsedId), keyItem.attributes);
        delete resultObject.attributes;
        if (keyItem.attributes.expires) {
            resultObject.expiresOn = keyItem.attributes.expires;
            delete resultObject.expires;
        }
        return resultObject;
    }
    /**
     * @internal
     * @ignore
     * Creates a span using the tracer that was set by the user.
     * @param {string} methodName The name of the method creating the span.
     * @param {RequestOptionsBase} [options] The options for the underlying HTTP request.
     */
    createSpan(methodName, requestOptions) {
        const tracer = getTracer();
        const span = tracer.startSpan(methodName, requestOptions && requestOptions.spanOptions);
        span.setAttribute("az.namespace", "Microsoft.KeyVault");
        return span;
    }
    /**
     * @internal
     * @ignore
     * Returns updated HTTP options with the given span as the parent of future spans,
     * if applicable.
     * @param {Span} span The span for the current operation.
     * @param {RequestOptionsBase} [options] The options for the underlying HTTP request.
     */
    setParentSpan(span, options = {}) {
        if (span.isRecording()) {
            const spanOptions = options.spanOptions || {};
            return Object.assign(Object.assign({}, options), { spanOptions: Object.assign(Object.assign({}, spanOptions), { parent: span.context(), attributes: Object.assign(Object.assign({}, spanOptions.attributes), { "az.namespace": "Microsoft.KeyVault" }) }) });
        }
        else {
            return options;
        }
    }
}
